Support Simple and Advanced Datums

keyan-m opened the pull request:
To allow further flexibility, this PR proposes the datum to be updated so that:
- The `owner` becomes optional, allowing a closed reclaim endpoint
- The routing fee is specified on a per UTxO basis instead of being hard-coded (minimum of 0)

keyan-m committed:
feat: optional owner and custom router fee

keyan-m committed:
feat: support receiver datum hash for minswap

mpetruska reviewed (approved):
:+1:, very nice job! Please update the documentation (README.md) as well.

keyan-m committed:
feat: validate min receive in minswap instance

keyan-m committed:
refactor: update readme and remove offchain

Given our dedicated off-chain repo, I've removed the Lucid example from
this repo to avoid further maintenance.

keyan-m committed:
chore: update compiled scripts

keyan-m committed:
feat: support simple and advanced datums

keyan-m commented:
As suggested by @solidsnakedev `SmartHandleDatum` is now a sum type, supporting either a `Simple` datum, where only an `owner` is stored, or an `Advanced` datum for more complex validator functions.

keyan-m committed:
fix: better fee reduction check

keyan-m committed:
feat: allow script context validation in adv. case

keyan-m committed:
fix: remove `Maybe` for `ScriptContext`

solidsnakedev commented:
LGTM

keyan-m committed:
feat: support advanced redeemers

keyan-m committed:
feat: remove advanced rdmrs, add advanced reclaim

Reverting advanced reclaims through routing logic proved infeasible as
the hardcoded swap address prevented a custom output.

With addition of `AdvancedReclaim` redeemer, any advanced datum with an
`owner` (i.e. `mOwner` being a `Just`) can be spent as long as
`validateFn` passes.

`validateFn` now also takes a `Bool` instead of an arbitrary `Redeemer`
to allow it distinguish between `Swap` or `Reclaim`.

keyan-m committed:
feat: extend advanced claim to batch version

keyan-m committed:
chore: rename "swap" to "route"

keyan-m committed:
feat: pass output value to custom validator

keyan-m committed:
fix: prevent deregistration

keyan-m committed:
fix: remove passing of output value to validator

keyan-m committed:
feat: pass router fee to the custom validator

keyan-m committed:
feat: pass input value to custom validator

keyan-m committed:
fix: batch reclaim

keyan-m committed:
feat: support user-approved reclaims for minswapv1

keyan-m committed:
chore: compile uplc

keyan-m committed:
refactor: `let` -> `plet`

keyan-m committed:
fix: remove `zw3rk` from nix flake

mpetruska commented on src/StakingValidator.hs:
```diff
@@ -51,6 +54,7 @@ data RouterRedeemer = RouterRedeemer
 PlutusTx.makeLift ''RouterRedeemer
 PlutusTx.makeIsDataIndexed ''RouterRedeemer [('RouterRedeemer, 0)]
 
+-- TODO: Switch to: '[ "indices" ':= PBuiltinList (PAsData (PBuiltinPair (PAsData PInteger) (PAsData PInteger)))
```
Is this a missed thing, or for the future?

keyan-m replied:
It's for future, to unify the interface with our Aiken's. I believe having a single list of pairs signals the underlying logic better. Although I'm not sure if it'd impact the performance negatively.

If having two lists is cheaper, perhaps we should refine the Aiken's interface?

mpetruska commented on src/Utils.hs:
```diff
@@ -140,8 +153,44 @@ pfoldl2 =
+              (pif (pnull # lc) acc perror)
+              lb
+        )
+        (pif (pnull # lb) acc perror)
```
```suggestion
        (pif (pand (pnull # lb) (pnull # lc)) acc perror)
```

mpetruska commented on src/StakingValidator.hs:
```diff
@@ -136,18 +169,61 @@ puniqueOrdered =
+smartHandleStakeValidatorW = phoistAcyclic $ plam $ \validateFn routeAddress redeemer ctx -> P.do
   let red = pconvertUnsafe @PRouterRedeemer redeemer
-  redF <- pletFields @'["inputIdxs", "outputIdxs"] red
+  redF <- pletFields @'["inputIdxs", "outputIdxs", "advancedRedeemers"] red
```
```suggestion
  redF <- pletFields @'["inputIdxs", "outputIdxs"] red
```

mpetruska commented on src/StakingValidator.hs:
```diff
@@ -136,18 +169,61 @@ puniqueOrdered =
   ctxF <- pletFields @'["txInfo", "purpose"] ctx
-  infoF <- pletFields @'["inputs", "outputs", "signatories", "datums"] ctxF.txInfo
+  PRewarding _ <- pmatch ctxF.purpose
+  infoF <- pletFields @'["inputs", "outputs", "signatories", "redeemers", "datums"] ctxF.txInfo
```
```suggestion
  infoF <- pletFields @'["inputs", "outputs", "redeemers", "datums"] ctxF.txInfo
```

mpetruska reviewed (changes requested):
Nice job, only minor changes requested.

keyan-m committed:
fix: remove `foldl3` & unnecessary field matches

keyan-m committed:
feat: require all mints to be reflected in reclaim

For the advanced reclaim endpoint, there is now a requirement to have
any mints occuring in the transaction to be included in the
corresponding output.

This allows burning of tokens the are being reclaimed.

keyan-m committed:
feat: implicit mint rqrmnt -> explicit datum field

keyan-m committed:
refactor: rename `ReclaimMint` -> `RequiredMint`

keyan-m committed:
Revert "rename `ReclaimMint` -> `RequiredMint`"

This reverts commit 47622a11c574f6c50bed833a8ed2aa1ac7ac0591.

keyan-m committed:
Revert "Revert "rename `ReclaimMint` -> `RequiredMint`""

This reverts commit f09da6a191762e69b663df3102e9f2cfd5772f35.

keyan-m committed:
feat: support `RequiredMint` for both adv. endpts

keyan-m committed:
chore: compile

keyan-m committed:
refactor: remove quantity from required mints

keyan-m committed:
chore: compile

keyan-m committed:
fix: consider ada in mint fields

keyan-m committed:
docs: update `README.md`

keyan-m committed:
docs: update table of contents

keyan-m committed:
docs: change title

keyan-m committed:
docs: add reference to associated off-chain tools

keyan-m committed:
chore: bump version

mpetruska commented on src/BatchValidator.hs:
```diff
@@ -38,23 +37,24 @@ instance PTryFrom PData PSmartRedeemer
 deriving via (DerivePConstantViaData SmartRedeemer PSmartRedeemer) instance PConstantDecl SmartRedeemer
 
+pstakeScriptIsInvoked :: Term s (PScriptContext :--> PStakingCredential :--> POpaque)
+pstakeScriptIsInvoked = plam $ \ctx stakeScript -> P.do
```
How about `passertStakeScriptIsInvoked`?

mpetruska commented on src/Utils.hs:
```diff
@@ -145,3 +204,125 @@ pconvertChecked x = ptryFrom x fst
+  Term s (PTriple PCurrencySymbol PTokenName PInteger)
+pgetSingleAssetApartFromADA v =
+  -- not using `pelimList` as it would've lead to evaluation of the head
+  psingleAssetToTriple #$ pheadSingleton $ ptail # presolveValueToList v
```
This only works for `'Sorted` key guarantee!

keyan-m replied:
Very nice catch! Thank you :pray:

mpetruska reviewed (changes requested):
Very nice work! I found only one issue.

keyan-m committed:
fix: add proper guarantee
